src/runtime/pkg/containerd-shim-v2/device_cold_plug.go

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2025 NVIDIA CORPORATION.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package containerdshim
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"runtime"
+
+	"github.com/kata-containers/kata-containers/src/runtime/pkg/device/config"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	podresourcesv1 "k8s.io/kubelet/pkg/apis/podresources/v1"
+)
+
+const (
+	nameAnnotation      = "io.kubernetes.cri.sandbox-name"
+	namespaceAnnotation = "io.kubernetes.cri.sandbox-namespace"
+)
+
+// coldPlugDevices handles cold plug of CDI devices into the sandbox
+// kubelet's PodResources API is used for determining the devices to be
+// cold plugged, if so configured. Otherwise, cdi annotations can be used for
+// covering stand alone scenarios.
+func coldPlugDevices(ctx context.Context, s *service, ociSpec *specs.Spec) error {
+	if !s.config.ColdPlugDevices {
+		// device cold plug is not enabled
+		return nil
+	}
+
+	kubeletSock := s.config.PodResourceAPISock
+	if kubeletSock != "" {
+		return coldPlugWithAPI(ctx, s, ociSpec)
+	}
+
+	// Here we deal with CDI devices that are cold-plugged
+	// for the single_container (nerdctl, podman, ...) use-case.
+	// We can provide additional directories where to search for
+	// CDI specs if needed. immutable OS's only have specific
+	// directories where applications can write too. For instance /opt/cdi
+	_, err := config.WithCDI(ociSpec.Annotations, []string{}, ociSpec)
```

Comment:
Now I wonder whether the output oci spec even makes sense, given that here we're apparently relying on the input being modified in-place.

Reply:
Sure, I was just thinking that you might already know the callers and be able to clean that up a bit. Related to https://github.com/kata-containers/kata-containers/pull/12087#discussion_r2526921283, too.